File: node_backend/src/services/user_service.ts
// services/user_service.ts — Equivalent to go_backend/services/user_service.go
import { UserDBOps } from '../dbops/user_dbops';
import { User } from '../entities/user';

/**
 * Request DTO for creating a user.
 * Mirrors Go's CreateUserRequest struct.
 */
export interface CreateUserRequest {
    name: string;
    email: string;
}

/**
 * Interface for User service operations.
 * Mirrors the Go UserService interface for dependency injection.
 */
export interface UserService {
    createUser(req: CreateUserRequest): Promise<User>;
    getUser(id: number): Promise<User>;
}

/**
 * Implementation of User service with business logic.
 * Equivalent to the Go userService struct.
 */
export class UserServiceImpl implements UserService {
    private userDBOps: UserDBOps;

    constructor(userDBOps: UserDBOps) {
        this.userDBOps = userDBOps;
    }

    async createUser(req: CreateUserRequest): Promise<User> {
        // Business logic: check if email already exists
        const existingUser = await this.userDBOps.getByEmail(req.email);
        if (existingUser) {
            throw new Error(`User with email ${req.email} already exists`);
        }

        const user = await this.userDBOps.create({
            name: req.name,
            email: req.email,
        });

        return user;
    }

    async getUser(id: number): Promise<User> {
        const user = await this.userDBOps.getByID(id);
        if (!user) {
            throw new Error(`Failed to get user: user not found`);
        }
        return user;
    }
}

File: go_backend/services/user_service.go
package services

import (
	"context"
	"fmt"

	"syncflow/dbops"
	"syncflow/entities"
)

type CreateUserRequest struct {
	Name  string `json:"name" binding:"required"`
	Email string `json:"email" binding:"required,email"`
}

type UserService interface {
	CreateUser(ctx context.Context, req CreateUserRequest) (*entities.User, error)
	GetUser(ctx context.Context, id uint) (*entities.User, error)
}

type userService struct {
	userDBOps dbops.UserDBOps
}

func NewUserService(userDBOps dbops.UserDBOps) UserService {
	return &userService{userDBOps: userDBOps}
}

func (s *userService) CreateUser(ctx context.Context, req CreateUserRequest) (*entities.User, error) {
	// Business logic: check if email already exists
	existingUser, _ := s.userDBOps.GetByEmail(ctx, req.Email)
	if existingUser != nil {
		return nil, fmt.Errorf("user with email %s already exists", req.Email)
	}

	user := &entities.User{
		Name:  req.Name,
		Email: req.Email,
	}

	if err := s.userDBOps.Create(ctx, user); err != nil {
		return nil, fmt.Errorf("failed to create user: %w", err)
	}

	return user, nil
}

func (s *userService) GetUser(ctx context.Context, id uint) (*entities.User, error) {
	user, err := s.userDBOps.GetByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("failed to get user: %w", err)
	}
	return user, nil
}
